File: Day 19/Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        n,right =len(s), len(p)

        if right > n:
            return []

        need = Counter(p)
        window = Counter(s[:right])
        start_index = []
        if need == window:
            start_index.append(0)
        for i in range(right, n):
            window[s[i]] += 1
            left_c = s[i- right]
            window[left_c] -=1
            if window[left_c] == 0:
                del window[left_c]
            if window == need:
                start_index.append(i - right + 1)
        return start_index

File: Day 19/Find All Anagrams in a String.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int n = s.size();
        int m = p.size();
        if (m > n) return {};

        vector<int> need(26, 0), window(26, 0);
        for (char c : p) need[c - 'a']++;

        vector<int> res;
        int left = 0, right = 0;

        while (right < n) {
            window[s[right] - 'a']++;
            right++;

            if (right - left == m) {
                if (window == need) res.push_back(left);

                window[s[left] - 'a']--;
                left++;
            }
        }
        return res;
    }
};
